docs/events/2023/docker-workshop.md

```diff
@@ -0,0 +1,235 @@
+
+Docker is an awesome technology that allows you to run your applications in containers, cross platform without installing them permanently on your system.
+
+In this workshop we went over how `Dockerfiles` and `docker-compose` files work as well as how to build your own images and run them.
+
+![Containers vs VM](../../assets/images/workshops/2023-docker/containers-vs-virtual-machines.jpg)
+
+Docker containers and virtual machines have a lot in common, they both allow you to create a template for your application and run it on any system. The main difference is that docker containers are much more lightweight and can be run on any system that has docker installed.
+
+Instead of packaging an entire operating system like a virtual machine, docker containers only package the dependencies that your application needs to run. This means that you can run multiple docker containers on the same system without them interfering with each other.
+
+## Prerequisites
+
+### Docker
+
+There are a number of ways to install Docker but the easiest way to install it is docker desktop. Docker desktop is a GUI application that help you manage your docker images and containers without having to use the command line.
+
+### Dive (Optional)
+
+Dive is a piece of software that allows you to see the result of different steps in your docker image. It is a great tool to help you debug your docker images and see what is taking up the most space.
+
+
+
+## `Dockerfile`s
+
+Docker files are the core of the software. They allow you to define the install steps of your software and the different things that it needs like ports and storage.
+
+This workshop uses this [Example Code](https://github.com/ProgSoc/DockerWorkshop2023) to demo how docker files and docker compose files work and includes two different projects that need different steps to install. One easy and one harder to show the different ways that you can install software.
+
+### Basic `Dockerfile` Structure
+
+```dockerfile
+# Use an official Node.js runtime as the base image
+FROM node:20-alpine
+
+# Set the working directory in the container
+WORKDIR /app
+
+# Copy the current directory contents into the container
+# This ignores files listed in .dockerignore, such as the node_modules folder
+COPY . .
+
+# Set our env
+ENV NODE_ENV production
+
+# Install the app dependencies
+RUN yarn install --production
+
+# Make port 3000 available to the outside
+EXPOSE 3000
+
+# Define the command to run the app
+CMD yarn start
+```
+
+The above file has a few different parts that are important to understand but in order to understand why they are so popular in industry we need to understand how docker sees Dockerfiles. 
+
+To use an analogy, if you wanted to make a cake and share how you made it with everyone you would write down the recipe. The recipe would have a list of ingredients and a list of steps to make the cake. Dockerfiles are the same, they are a list of ingredients and steps to make your application.
```

Comment:
I feel like this analogy isn't great, it doesn't explain how docker layers still keep deleted files from previous layers. I think a better analogy is transparent overhead projector sheets (include image).

```
Each line in the file above creates a new "layer", which is like an addition on top of the previous image. Remember the old overhead projectors used in classrooms? Think of Docker layers like those transparent sheets. Each sheet has some information, and when stacked upon one another, they form a complete picture or presentation. You can add extra sheets on top to build up on your image, but if one sheet's colour overlaps with another, the bottom sheet is still there taking up space even if it's no longer visible.

Each set of Docker layers is represented by a hash, so for example when you download 2 Docker images and they both share the same core layers (e.g. both are built on top of the `node:20-alpine` Docker image), then those core layers are only downloaded once.
```
something along these lines.